File: 프로그래머스/unrated/132265. 롤케이크 자르기/롤케이크 자르기.py
def solution(topping):
    answer = 0
    dic = {}
    for t in topping:
        if t not in dic:
            dic[t] = 1
        else:
            dic[t] += 1
    
    dic_len = len(dic)
    b = {}
    b_len = 0
    for t in topping:
        dic[t] -= 1
        if dic[t] == 0:
            dic_len -= 1
        if t not in b:
            b[t] = 1
            b_len += 1
        
        if b_len == dic_len:
            answer += 1
            
    return answer
           

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.*;

class Solution {
    public int solution(int[] topping) {
        int cnt = 0;
        Map<Integer, Integer> dic = new HashMap<>();

        for (Integer t : topping) {
            if (dic.containsKey(t)) {
                dic.put(t, dic.get(t) + 1);
            } else {
                dic.put(t, 1);
                cnt += 1;
            }
        }

        Set<Integer> set = new HashSet<>();
        int answer = 0;
        for (Integer t : topping) {
            set.add(t);
            int v = dic.get(t) - 1;
            dic.put(t, v);
            if (v == 0) {
                cnt -= 1;
            }

            if (set.size() == cnt) {
                answer += 1;
            }
        }

        return answer;
    }
}